### ISSUE TITLE
Perl_sync_locale stopped working somewhere after 5.37

### ISSUE THREAD

**Issue by dk** [Perl/perl5#20565]
I'm having a problem in my XS code that sync_locale() stopped working in 5.37.5 -
the issue is similar to what was already once talked about back in 2019 ( https://rt.perl.org/Ticket/Display.html?id=133945 )

Please find attached a proper test case, but shortly,  it happens under LC_NUMERIC=de_DE.UTF-8 when C code
```
    sync_locale();
    SvNV(newSVpv("4.8888", 0))
```
returns 4.0 .

Important: In order to reproduce the bug your system needs to have the de_DE.UTF-8 locale installed.

**Comment by dk** [comment]
[sync_locale_bug.zip](https://github.com/Perl/perl5/files/10127115/sync_locale_bug.zip)

unpack, then perl Makefile.PL, make, ./test

**Comment by jkeenan** [comment]
> I'm having a problem in my XS code that sync_locale() stopped working in 5.37.5 

Can you provide the configuration arguments you used to build the `perl` executable at which you observe this problem?  Perhaps `./perl -Ilib -V:config_args`

You say this stopped working in 5.37.5.  Does that mean that it was working in 5.37.4?

**Comment by jkeenan** [comment]
I am having trouble reproducing your failure.
```
$ uname -mrs
FreeBSD 12.3-RELEASE-p1 amd64

$ locale -a | grep -E 'de_DE\.UTF-8'
de_DE.UTF-8

$ export THISDIR="$TESTINGDIR/v5.37.6"
$ export THISPERL="$THISDIR/bin/perl -I$THISDIR/lib"
$ $THISPERL -v | head -2 | tail -1
This is perl 5, version 37, subversion 6 (v5.37.6) built for amd64-freebsd-thread-multi
$ $THISPERL Makefile.PL && make
v5.37.6
Generating a Unix-style Makefile
Writing Makefile for T
Writing MYMETA.yml and MYMETA.json
Skip blib/lib/T.pm (unchanged)
Running Mkbootstrap for T ()
chmod 644 "T.bs"
"/usr/home/jkeenan/testing/v5.37.6/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- T.bs blib/arch/auto/T/T.bs 644
cc -c  -I/usr/local/include/gtk-3.0 -I/usr/local/include/pango-1.0 -I/usr/local/include -I/usr/local/include/glib-2.0 -I/usr/local/lib/glib-2.0/include -I/usr/local/include/harfbuzz -I/usr/local/include/freetype2 -I/usr/local/include/libpng16 -I/usr/local/include/fribidi -I/usr/local/include/cairo -I/usr/local/include/pixman-1 -I/usr/local/include/gdk-pixbuf-2.0 -I/usr/local/include/gio-unix-2.0 -I/usr/local/include/libepoll-shim -I/usr/local/include/atk-1.0 -I/usr/local/include/at-spi2-atk/2.0 -I/usr/local/include/dbus-1.0 -I/usr/local/lib/dbus-1.0/include -I/usr/local/include/at-spi-2.0 -D_THREAD_SAFE -pthread  -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -O2 -pipe -fstack-protector -fno-strict-aliasing    -DVERSION=\"\"  -DXS_VERSION=\"\" -DPIC -fPIC "-I/home/jkeenan/testing/v5.37.6/lib/perl5/5.37.6/amd64-freebsd-thread-multi/CORE"   T.c
rm -f blib/arch/auto/T/T.so
LD_RUN_PATH="/usr/local/lib" cc  -shared  -L/usr/local/lib -fstack-protector-strong  T.o  -o blib/arch/auto/T/T.so    -L/usr/local/lib -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lpthread -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lintl    
chmod 755 blib/arch/auto/T/T.so
```
Since your `./test` is a shell program, I have to hack it up a bit to use the `perl-5.37.6` described above instead of the system `perl`.
```
$ cat ./test
#!/bin/sh

echo "no bug expected:"
$THISPERL -v | head -2 | tail -1
$THISPERL -V:config_args
$THISPERL test.pl
export LC_NUMERIC=de_DE.UTF-8
echo "bug expected:"
$THISPERL test.pl
```
Running the test:
```
$ ./test
no bug expected:
This is perl 5, version 37, subversion 6 (v5.37.6) built for amd64-freebsd-thread-multi
config_args='-des -Dusedevel -Uversiononly -Dprefix=/home/jkeenan/testing/v5.37.6 -Dman1dir=none -Dman3dir=none -Duseithreads -Doptimize=-O2 -pipe -fstack-protector -fno-strict-aliasing';
no bug
bug expected:
bug
```

**Comment by dk** [comment]
James, the very last line of your output says "bug" -- this is it, you have it reproduced

**Comment by dk** [comment]
As for 5.37.4 I cannot say whether it was there or not, - I was going to go down the rabbit hole of git bisecting, but thankfully Karl Williamson kindly agreed to look up for the fix without me doing that. 5.37.5 is just a point of reference from cpantesters report.

**Comment by dk** [comment]
@khwilliamson

**Comment by andk** [comment]
Just for the record in case it helps somebody: I have bisected a close cousin of the given test harness here, KARASIK/Prima-1.67.tar.gz with the locale de_DE.utf8 and reached v5.37.4-55-g9f5a615be6, i.e.
```
commit 9f5a615be674d7663d3b4719849baa1ba3027f5b
Author: Karl Williamson <khw@cpan.org>
Date:   Sun Sep 11 18:08:32 2022 -0600

    locale.c: Revamp sync_locale(), switch_to_global_locale()
```

**Comment by jkeenan** [comment]
In https://github.com/Perl/perl5/issues/20617, Andreas (@andk) identified commit 818cdb7aa9f85227c1c7313257c6204c872beb94 as the commit which triggered a test failure in a CPAN module run against a `perl` built with locale de_DE.utf8.

My manual bisection indicates that 818cdb7aa9f is also the commit responsible for the problem reported in *this* issue.

I built a `perl` with my usual configuration options on FreeBSD-12 at the "before" commit.
```
$ ./bin/perl -v | head -2 | tail -1
This is perl 5, version 37, subversion 4 (v5.37.4 (v5.37.3-159-g3d3a0c7eea)) built for amd64-freebsd-thread-multi
$ export THISDIR=$TESTINGDIR/3d3a0c7eea7bfaa5d3cfad5522675a66d363afc7
$ export THISPERL="$THISDIR/bin/perl -I$THISDIR/lib"
$ $THISPERL Makefile.PL && make && ./test ; echo $?
...
<> no bug expected:
This is perl 5, version 37, subversion 4 (v5.37.4 (v5.37.3-159-g3d3a0c7eea)) built for amd64-freebsd-thread-multi
Got: 4.8888
no bug
<de_DE.UTF-8> bug expected:
Got: 4.8888
no bug
0
```
I next built a `perl` with my usual configuration options on FreeBSD-12 at the "after" commit.
```
$ ./bin/perl -v | head -2 | tail -1
This is perl 5, version 37, subversion 4 (v5.37.4 (v5.37.3-160-g818cdb7aa9)) built for amd64-freebsd-thread-multi
$ export THISDIR=/home/jkeenan/testing/818cdb7aa9
$ export THISPERL="$THISDIR/bin/perl -I$THISDIR/lib"
$ $THISPERL Makefile.PL && make && ./test ; echo $?
...
<> no bug expected:
This is perl 5, version 37, subversion 4 (v5.37.4 (v5.37.3-160-g818cdb7aa9)) built for amd64-freebsd-thread-multi
Got: 4.8888
no bug
<de_DE.UTF-8> bug expected:
Got: 4
bug
1
```
For the above, to avoid making the same mistake I made earlier in this ticket, I modified the OP's scripts as follows:
```
$ cat ./test.pl
use strict;
use warnings;
use lib 'blib/arch', '.';
use T;
my $got = T::alpha();
print "Got: $got\n";
#print (((4 == T::test()) ? "bug" : "no bug"), "\n");
#print (((4 == $got) ? "bug" : "no bug"), "\n");
if (4 == $got) {
    print "bug\n";
    exit 1;
}
else {
    print "no bug\n";
    exit 0;
}
```
```
$ cat ./test
#!/bin/sh

echo "<$LC_NUMERIC> no bug expected:"
#perl test.pl
$THISPERL -v | head -2 | tail -1
$THISPERL test.pl
export LC_NUMERIC=de_DE.UTF-8
echo "<$LC_NUMERIC> bug expected:"
#perl test.pl
$THISPERL test.pl
exit $?
```
So the problematic commit here appears to be:
```
commit 818cdb7aa9f85227c1c7313257c6204c872beb94 (HEAD)
Author:     Karl Williamson <khw@cpan.org>
AuthorDate: Sun Apr 11 05:57:07 2021 -0600
Commit:     Karl Williamson <khw@cpan.org>
CommitDate: Thu Sep 1 09:02:04 2022 -0600

    locale.c: Skip code if will be a no-op
    
    The previous commits have fixed things up so that at this point in the
    code nothing has changed, and if nothing will change, we can just return
```
@khwilliamson, can you take a look?

Thank you very much.

**Comment by jkeenan** [comment]
This problem appears to be resolved with changes @khwilliamson proposes in the smoke-me/khw-dimitry branch (https://github.com/Perl/perl5/pull/20618).
```
$ ./bin/perl -Ilib -v | head -2 | tail -1
This is perl 5, version 37, subversion 7 (v5.37.7 (v5.37.6-107-ga16e29ec4a)) built for amd64-freebsd-thread-multi

$ ./bin/perl -Ilib -V:config_args
config_args='-des -Dusedevel -Uversiononly -Dprefix=/home/jkeenan/testing/smoke-me/khw-dimitry -Dman1dir=none -Dman3dir=none -Duseithreads -Doptimize=-O2 -pipe -fstack-protector -fno-strict-aliasing';

$ export THISDIR=/home/jkeenan/testing/smoke-me/khw-dimitry
$ export THISPERL="$THISDIR/bin/perl -I$THISDIR/lib"
$ $THISPERL Makefile.PL && make && ./test ; echo $?
...
<> no bug expected:
This is perl 5, version 37, subversion 7 (v5.37.7 (v5.37.6-107-ga16e29ec4a)) built for amd64-freebsd-thread-multi
Got: 4.8888
no bug
<de_DE.UTF-8> bug expected:
Got: 4.8888
no bug
0
```

**Comment by khwilliamson** [comment]
Fix now released in 5.37.7

**Comment by khwilliamson** [comment]
by https://github.com/Perl/perl5/commit/cb777d7125776d73756033f32e5e2040a8826cc1
